narwhal/consensus/src/bullshark.rs

```diff
@@ -158,39 +153,29 @@ impl ConsensusProtocol for Bullshark {
 
             // Starting from the oldest leader, flatten the sub-dag referenced by the leader.
             for x in utils::order_dag(self.gc_depth, leader, state) {
-                let digest = x.digest();
-
                 // Update and clean up internal state.
                 state.update(&x, self.gc_depth);
 
                 // Add the certificate to the sequence.
-                sequence.push(ConsensusOutput {
-                    certificate: x,
-                    consensus_index,
-                });
-
-                // Increase the global consensus index.
-                consensus_index += 1;
-
-                // Persist the update.
-                // TODO [issue #116]: Ensure this is not a performance bottleneck.
-                self.store.write_consensus_state(
-                    &state.last_committed,
-                    &consensus_index,
-                    &digest,
-                )?;
-                debug!(
-                    "Store commit index:{}, digest:{}",
-                    &consensus_index, &digest
-                );
+                sequence.push(x);
             }
 
+            let next_sub_dag_index = state.latest_sub_dag_index + 1;
             let sub_dag = CommittedSubDag {
                 certificates: sequence,
                 leader: leader.clone(),
+                sub_dag_index: next_sub_dag_index,
             };
+
+            // Increase the global consensus index.
+            state.latest_sub_dag_index = next_sub_dag_index;
+
+            // Persist the update.
+            // TODO [issue #116]: Ensure this is not a performance bottleneck.
             self.store
-                .write_committed_sub_dag(&state.last_committed, &sub_dag)?;
+                .write_consensus_state(&state.last_committed, &sub_dag)?;
```

Comment:
Because you have `latest_sub_dag_index` in memory, you should maintain consistency between what is in DB and in memory. This consistency can be violated if method here returns error. (in memory state is updated and db is not)

You can fix it just by swapping two lines and updating in-memory state after db write succeed

Reply:
good catch